### ISSUE TITLE
Shader + Matrix: Matrix3 and Matrix4 Uniform uploading

### ISSUE THREAD

**Issue by OkiStuff** [PikoStudios/Muzzle#34]
# Shader: Matrix3 and Matrix4 Uniform uploading
Muzzle should have a `upload_shader_matrix3()` and a `upload_shader_matrix4()` function in `modern_pipeline/Shaders.h`

## Matrix
A new `Matrix.h` should be creating containing Matrix3 and Matrix4 types.
```c
#pragma once

typedef struct matrix3
{
...
} matrix3;

typedef struct matrix4
{
...
} matrix4;
```

## Branch
All PRs for this issue should be made to the `revamped_renderer` branch

**Comment by Vaimer9** [comment]
on it

**Comment by Vaimer9** [comment]
In `Shader.h` you have provided a reference snippet from raylib. They've impled like so -
```c
typedef struct Matrix {
    float m0, m4, m8, m12;  // Matrix first row (4 components)
    float m1, m5, m9, m13;  // Matrix second row (4 components)
    float m2, m6, m10, m14; // Matrix third row (4 components)
    float m3, m7, m11, m15; // Matrix fourth row (4 components)
} Matrix;
```
I was thinking more of a 1 dimensional array with length 9/16 respectively for mat3/4.

Raylib's implementation prevents the user from using a smaller sized array is all. Would like some feedback.

**Comment by Vaimer9** [comment]
Also, just putting this out there -

This API *will* be deprecated at some point. `cglm` uses SIMD for a most computations which is very useful. Not to mention the tedious effort which has to be put in for all of the matrix operations if this is maintained by Muzzle.

**Comment by OkiStuff** [comment]
> In `Shader.h` you have provided a reference snippet from raylib. They've impled like so -
> 
> ```c
> typedef struct Matrix {
>     float m0, m4, m8, m12;  // Matrix first row (4 components)
>     float m1, m5, m9, m13;  // Matrix second row (4 components)
>     float m2, m6, m10, m14; // Matrix third row (4 components)
>     float m3, m7, m11, m15; // Matrix fourth row (4 components)
> } Matrix;
> ```
> 
> I was thinking more of a 1 dimensional array with length 9/16 respectively for mat3/4.
> 
> ```c
> typedef float* mat3;
> typedef float* mat4;
> // so on for ints
> ```
> 
> Raylib's implementation prevents the user from using a smaller sized array is all. Would like some feedback.

I agree with you, as OpenGL also takes in a 1D array.

**Comment by OkiStuff** [comment]
> Also, just putting this out there -
> 
> This API _will_ be deprecated at some point. `cglm` uses SIMD for a most computations which is very useful. Not to mention the tedious effort which has to be put in for all of the matrix operations if this is maintained by Muzzle team.

I do not intend to make any extra support for Matrices in MuzzleMath, if someone requests it, I will look into `cglm`. Unless you think it's important to have Matrix Operations in Muzzle.

**Comment by OkiStuff** [comment]
This issue has been resolved by PR #36. Please head to #35 for CGLM discussion.
